utils/pre-swap-checker/contracts/PreSwapChecker.sol

```diff
@@ -26,69 +30,126 @@ contract PreSwapChecker {
 
     // Check self transfer
     if (order.signer.wallet == order.sender.wallet) {
-      errors[errorCount] = 'SELF_TRANSFER_INVALID';
+      errors[errorCount] = "SELF_TRANSFER_INVALID";
       errorCount++;
     }
 
     // Check expiry
     if (order.expiry < block.timestamp) {
-      errors[errorCount] = 'ORDER_EXPIRED';
+      errors[errorCount] = "ORDER_EXPIRED";
       errorCount++;
     }
 
     if (ISwap(swap).signerNonceStatus(order.signer.wallet, order.nonce) != 0x00) {
-      errors[errorCount] = 'ORDER_TAKEN_OR_CANCELLED';
+      errors[errorCount] = "ORDER_TAKEN_OR_CANCELLED";
       errorCount++;
     }
 
     if (order.nonce < ISwap(swap).signerMinimumNonce(order.signer.wallet)) {
-      errors[errorCount] = 'NONCE_TOO_LOW';
+      errors[errorCount] = "NONCE_TOO_LOW";
       errorCount++;
     }
 
-    // Check the order sender
-    if (order.sender.wallet != address(0)) {
-      // The sender was specified
-      // Check their token balance
-      if (!hasBalance(order.sender)) {
-        errors[errorCount] = 'SENDER_BALANCE';
-        errorCount++;
-      }
+    // check if ERC721 or ERC20 only amount or id set for sender
+    if (order.sender.kind == ERC20_INTERFACE_ID && order.sender.id != 0) {
+      errors[errorCount] = "SENDER_INVALID_ID";
+      errorCount++;
+    } else if (order.sender.kind == ERC721_INTERFACE_ID && order.sender.amount != 0) {
+      errors[errorCount] = "SENDER_INVALID_AMOUNT";
+      errorCount++;
+    }
 
-      // Check their approval
-      if (!isApproved(order.sender, swap)) {
-        errors[errorCount] = 'SENDER_ALLOWANCE';
-        errorCount++;
-      }
+    // check if ERC721 or ERC20 only amount or id set for signer
+    if (order.signer.kind == ERC20_INTERFACE_ID && order.signer.id != 0) {
+      errors[errorCount] = "SIGNER_INVALID_ID";
+      errorCount++;
+    } else if (order.signer.kind == ERC721_INTERFACE_ID && order.signer.amount != 0) {
+      errors[errorCount] = "SIGNER_INVALID_AMOUNT";
+      errorCount++;
     }
 
-    // Check the order signer
-    if (!hasBalance(order.signer)) {
-      errors[errorCount] = 'SIGNER_BALANCE';
+    // Check valid token registry handler for sender
+    if (hasValidKind(order.sender.kind, swap)) {
+      // Check the order sender
+      if (order.sender.wallet != address(0)) {
+        // The sender was specified
+        // Check if sender kind interface can correctly check balance
+        if (order.sender.kind == ERC721_INTERFACE_ID && !hasValidERC71Interface(order.sender)) {
```

Comment:
the function `hasValidERC721Interface` already does the check that `party.kind == ERC721_INTERFACE_ID`, do that check doesnt need to be in both places